libsql/src/hrana/stream.rs

```diff
@@ -195,25 +209,58 @@ where
 #[derive(Debug)]
 struct RawStream<T>
 where
-    T: for<'a> HttpSend<'a>,
+    T: HttpSend,
 {
     client: T,
     baton: Option<String>,
-    base_url: String,
-    auth_token: String,
+    pipeline_url: Arc<str>,
+    auth_token: Arc<str>,
     status: StreamStatus,
     sql_id_generator: SqlId,
+    cursor_url: Arc<str>,
 }
 
 impl<T> RawStream<T>
 where
-    T: for<'a> HttpSend<'a>,
+    T: HttpSend,
 {
     async fn send(&mut self, req: StreamRequest) -> Result<StreamResponse> {
         let [resp] = self.send_requests([req]).await?;
         Ok(resp)
     }
 
+    pub async fn open_cursor(&mut self, batch: Batch) -> Result<Cursor<T::Stream>> {
+        if self.status == StreamStatus::Closed {
+            return Err(HranaError::StreamClosed(
+                "client stream has been closed by the servers".to_string(),
+            ));
+        }
+        let msg = CursorReq {
+            baton: self.baton.clone(),
+            batch,
+        };
+        let body = serde_json::to_string(&msg).map_err(HranaError::Json)?;
+        let stream = self
+            .client
+            .http_send(self.cursor_url.clone(), self.auth_token.clone(), body)
+            .await?;
+        let (cursor, mut response) = Cursor::open(stream).await?;
+        if let Some(base_url) = response.base_url.take() {
+            self.cursor_url = Arc::from(base_url);
```

Comment:
What is the case that triggers this?

Reply:
From Hrana spec: https://github.com/tursodatabase/libsql/blob/main/docs/HRANA_3_SPEC.md

> If the base_url in the response is not null, the client should use this URL when sending further requests on this stream. If it is null, the client should use the same URL that it has used for the previous request. The base_url must be an absolute URL with "http" or "https" scheme.

Reply:
I've changed this part - it looks like `base_url` needs manual path suffix addition for `/v3/pipeline` and `/v3/cursor`. The reference from typescript client: https://github.com/libsql/hrana-client-ts/blob/5db07653c14c26afe5f0b9039716aad47753420c/src/http/stream.ts#L348